### ISSUE TITLE
RandomWalk much slower when sheared

### ISSUE THREAD

**Issue by esheldon** [GalSim-developers/GalSim#977]
The following code runs 10 times faster when no shear is applied.
```python
import numpy as np
import galsim
import time

import argparse

parser=argparse.ArgumentParser()
parser.add_argument('ntrial',type=int)
parser.add_argument('--shear',action='store_true')


class Sim(object):
    def __init__(self, shear=False):
        self.shear=shear
        self.dim=100

        self.flux = 100.0
        self.hlr = 4.0
        self.n_knots=100
        self.psf=galsim.Gaussian(fwhm=0.9)

    def make_im(self):
        obj=self.get_model()
        if self.shear:
            g1,g2=self.get_shape()
            obj = obj.shear(g1=g1,g2=g2)

        obj=galsim.Convolve(obj, self.psf)
        return obj.drawImage(
            scale=0.263,
        ).array

    def get_shape(self):
        while True:
            g1,g2 = np.random.normal(scale=0.2,size=2)
            g=np.sqrt(g1**2 + g2**2)
            if g < 0.99:
                break
        return g1,g2

    def get_model(self):
        profile=galsim.Exponential(
            flux=self.flux,
            half_light_radius=self.hlr,
        )

        obj=galsim.RandomWalk(
            self.n_knots,
            profile=profile,
        )
        return obj

args=parser.parse_args()
sim=Sim(shear=args.shear)

tm=time.time()
for i in range(args.ntrial):
    im=sim.make_im()

print("time:",time.time()-tm)
```

The profiler indicates that it is spending a lot of time in transform.py  Is this something we could optimize?  This is pure peculation, but these are point sources (a delta function is used for each knot) so we should be able to do something smart to avoid this overhead it seems.

```
time: 14.920971870422363
         1162298 function calls (1135947 primitive calls) in 18.439 seconds

   Ordered by: internal time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
      100   14.072    0.141   14.099    0.141 transform.py:500(_drawKImage)
100342/100340    0.171    0.000    0.261    0.000 {built-in method builtins.isinstance}
      100    0.156    0.002    0.196    0.002 randwalk.py:197(_sbp)
      100    0.094    0.001    0.094    0.001 {built-in method galsim._galsim.irfft}
 1388/259    0.088    0.000    0.327    0.001 sre_parse.py:470(_parse)
2804/2769    0.084    0.000    0.602    0.000 core.py:2031(_expand_and_gather)
      531    0.083    0.000    0.083    0.000 {built-in method marshal.loads}
... snip ...
```

**Comment by rmjarvis** [comment]
To be fair, the `_drawKImage` called from transform.py is the starting point for doing all the drawing.  Including the delta functions.  But yes, there is probably a fair amount of inefficiency in that, since it's trying to be general and doesn't know the next layer down is just a sum of delta functions.

My recommendation would be to override the `shear`, `transform`, etc. methods in `RandomWalk` to just shift the positions of the delta functions around.  Or in particular transposing the order of operations of the `shoot` part and the transformation.  If you shoot the transposed profile, the knots will come out in the right place.

So something like this:
```
def shear(self, *args, **kwargs):
    return RandomWalk(self.npoints, profile=self.profile.shear(*args, **kwargs), ...)
```
And likewise for the other transformations.

**Comment by esheldon** [comment]
makes sense, thanks.  When I get a chance I'll do a PR

**Comment by esheldon** [comment]
this doesn't work as expected because it makes a new object and new points. For example if I wanted to generate 3 versions with different flux, it would make new points also rather than just the same points with different flux.

I think we would have to add another feature that the points can also be sent on the constructor

**Comment by rmjarvis** [comment]
Actually, this points out that `__repr__` and `__eq__` are also incorrect currently.  `eval(repr(rw))` currently returns a new realization with different points.  And `__eq__` erroneously declares them to be equal.

So along the way when you're working on this, you should add the test:
```
do_pickle(conv, lambda x: x.drawImage())
```
which currently fails.  The regular pickle test uses `__eq__` which wrongly thinks the two objects are equivalent.

I think `__eq__` should check that the positions of the points are equal.  And for  `__repr__`, I can think of two options.
1. You can use a constructor argument that explicitly lists the point positions rather than generate new ones.
2. You can save the rng state (with `rng.duplicate()`) before generating the point positions and then pass that as the random number generator to use in `__repr__`.

Either of these would also facilitate the `shear` override to preserve the point positions.  The second one would likely be easier, since you can rely on the shoot function still to generate the sheared positions.  In option 1, you would probably need to move the points by hand.

**Comment by rmjarvis** [comment]
Fixed via #1051
